src/app/services/todos-store.service.ts

```diff
@@ -0,0 +1,79 @@
+import { Injectable } from '@angular/core';
+import { BehaviorSubject } from 'rxjs';
+import { shareReplay, map } from 'rxjs/operators';
+
+import Todo from 'src/app/models/todo.model';
+import ITodo from 'src/app/models/itodo';
+import { TodosService } from 'src/app/services/todos.service';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class TodosStoreService {
+  private readonly _todos = new BehaviorSubject<ITodo[]>([]);
+
+  readonly todos$ = this._todos.asObservable();
+
+  get todos(): ITodo[] {
+    return this._todos.getValue();
+  }
+  set todos(todos: ITodo[]) {
+    this._todos.next(todos);
+  }
+
+  constructor(private todosService: TodosService) {
+    this.fetchAll();
+  }
+
+  async fetchAll() {
+    this.todos = await this.todosService.getAll().toPromise();
+  }
+
+  async add(title: string) {
+    const optimisticTodo = new Todo(title);
+    this.todos = [...this.todos, optimisticTodo];
+
+    try {
+      const todo = await this.todosService.create(optimisticTodo).toPromise();
+
+      const index = this.todos.indexOf(
+        this.todos.find((t) => t.id === optimisticTodo.id)
+      );
+
+      this.todos[index] = todo;
+    } catch (e) {
+      // if server returns an error -> revert the changes
+      console.error(e);
+      this.remove(optimisticTodo.id, false);
+    }
+  }
+
+  async remove(id: string, serverRemove = true) {
+    const todo = this.todos.find((t) => t.id === id);
+    this.todos = this.todos.filter((t) => t.id !== id);
```

Comment:
if this method is amended in the same way as `add()`, then `todos` should be filtered only in successful `next` call. Thus, it would not be needed to restore removed `todo`

Reply:
Regarding 
> Thus, it would not be needed to restore removed todo

I wanted to create "optimistic" CRUD actions on UI side, so that if user tries to add, delete, etc. a Todo, they should **immediately see the result** of the action, while server processes the query.

And if server's response has an error, we reroll front back to what it should be due to errors on BE side